File: Gaussian/Cpp/solve_system.cpp
/*!
    \file   solve_system.c
    \brief  Solve a large system of simultaneous equations.
    \author (C) Copyright 2011 by Peter C. Chapin <PChapin@vtc.vsc.edu>
*/

#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <fstream>
#include <Timer.hpp>

// Select the serial or parallel version as desired...
// #include "linear_equations.hpp"
#include "linear_equationsp.hpp"

using namespace std;

int main( int argc, char *argv[] )
{
    size_t size;

    if( argc != 2 ) {
        cout << "Error: Expected the name of a system definition file.\n";
        return EXIT_FAILURE;
    }

    ifstream input_file( argv[1] );
    if( !input_file ) {
        printf("Error: Can not open the system definition file.\n");
        return EXIT_FAILURE;
    }

    // Get the size.
    input_file >> size;

    // Allocate the arrays.
    Matrix<float> a( size, size );
    boost::scoped_array<float> b( new float[size] );

    // Get coefficients.
    for( size_t i = 0; i < size; ++i ) {
        for( size_t j = 0; j < size; ++j ) {
            input_file >> a(i, j);
        }
        input_file >> b[i];
    }
    input_file.close( );

    spica::Timer stopwatch;
    stopwatch.start( );
    bool success = gaussian_solve( a, b.get( ) );
    stopwatch.stop( );

    if( !success ) {
        cout << "System is degenerate\n";
    }
    else {
        cout << "\nSolution is\n";
        for( size_t i = 0; i < size; ++i ) {
            cout << " x(" << setw(4) << i << ") = "
                 << setw(9) << std::fixed << std::setprecision(5) << b[i] << "\n";
        }

        cout << "\nExecution time = " << stopwatch.time( ) << " milliseconds\n";
    }

    return EXIT_SUCCESS;
}

File: Gaussian/Fortran/solve_system.f90
!===========================================================================
! FILE          : solve_system.f90
! LAST REVISION : 2015-04-06
! SUBJECT       : Example program that solves simultaneous equations.
!
! This program is a lightly modified version of a program that is shown as an example in
! "Fortran 90 Programming" by Ellis, Philips, and Lahey. Published by Addison Wesley, (C)
! Copyright 1994, ISBN=0-201-54446-6. Section 18.5, page 647.
!
! Please send comments or bug reports to
!
!      Peter C. Chapin
!      Computer Information Systems Department
!      Vermont Technical College
!      Williston, VT 05495
!      pchapin@vtc.edu
!===========================================================================

PROGRAM solve_system
  USE linear_equations
  IMPLICIT NONE

  ! Allocatable arrays for coefficients
  REAL*8, ALLOCATABLE, DIMENSION(:, :) :: a
  REAL*8, ALLOCATABLE, DIMENSION(:)    :: b

  ! Size of arrays
  INTEGER :: n

  ! Loop variables and error flag
  INTEGER :: i, j, error

  ! Timing information
  REAL :: start_time, stop_time

  ! Get size of problem
  READ *, n

  ! Allocate arrays
  ALLOCATE(a(n, n), b(n))

  ! Get coefficients
  DO i = 1, n
     READ *, (a(j, i), j = 1, n), b(i)
  END DO

  PRINT *, "Coefficients read"

  ! Solve the system
  CALL CPU_TIME(start_time)
  CALL gaussian_solve(a, b, n, error)
  CALL CPU_TIME(stop_time)

  ! Did it work?
  IF (error == -1) THEN
     PRINT *, "Error in call to gaussian_solve"
  ELSE IF (error == -2) THEN
     PRINT *, "System is degenerate"
  ELSE
     PRINT *, " "
     PRINT *, "Solution is"
     PRINT '(1X,"x(",I4,") = ",F9.5)', (i, b(i), i = 1, n)
     PRINT *, 'Time = ', stop_time - start_time, ' seconds.'
  END IF

END PROGRAM solve_system
